File: server/internal/domain/lending_test.go
package domain_test

import (
	"errors"
	"testing"
	"time"

	"github.com/google/uuid"
	"github.com/kotarowakuwaku/asset-wish/server/internal/domain"
)

func newLending(t *testing.T, amount, collected domain.Money) domain.Lending {
	t.Helper()
	l, err := domain.NewLending(uuid.New(), "田中", "", amount, time.Date(2026, 7, 12, 0, 0, 0, 0, time.UTC))
	if err != nil {
		t.Fatalf("NewLending: %v", err)
	}
	l.CollectedAmount = collected
	return l
}

func TestLending_Collect(t *testing.T) {
	tests := []struct {
		name          string
		amount        domain.Money
		collected     domain.Money
		collectAmount domain.Money
		wantErr       error
		wantCollected domain.Money
		wantStatus    domain.CollectionStatus
	}{
		{"F-1: partial collect", 12000, 0, 5000, nil, 5000, domain.CollectionPartial},
		{"F-2: reach full", 12000, 5000, 7000, nil, 12000, domain.CollectionCollected},
		{"F-3: over collect (already partial)", 12000, 5000, 8000, domain.ErrCollectExceedsOutstanding, 5000, domain.CollectionPartial},
		{"F-4: over collect (already full)", 12000, 12000, 1, domain.ErrCollectExceedsOutstanding, 12000, domain.CollectionCollected},
		{"F-5: zero amount", 12000, 0, 0, domain.ErrInvalidAmount, 0, domain.CollectionUncollected},
		{"F-6: negative amount", 12000, 0, -100, domain.ErrInvalidAmount, 0, domain.CollectionUncollected},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			l := newLending(t, tt.amount, tt.collected)
			err := l.Collect(tt.collectAmount)
			if !errors.Is(err, tt.wantErr) {
				t.Errorf("err=%v want %v", err, tt.wantErr)
			}
			if l.CollectedAmount != tt.wantCollected {
				t.Errorf("CollectedAmount=%d want %d", l.CollectedAmount, tt.wantCollected)
			}
			if l.Status() != tt.wantStatus {
				t.Errorf("Status()=%s want %s", l.Status(), tt.wantStatus)
			}
		})
	}
}

func TestLending_OutstandingAndStatus(t *testing.T) {
	l := newLending(t, 12000, 0)
	if got := l.Outstanding(); got != 12000 {
		t.Errorf("F-7: Outstanding()=%d want 12000", got)
	}
	if got := l.Status(); got != domain.CollectionUncollected {
		t.Errorf("F-7: Status()=%s want uncollected", got)
	}
}

File: worker/src/domain/lending.test.ts
import { describe, expect, it } from 'vitest'
import { expectDomainError, id, lend, SOME_DATE, yen } from '../../test/support'
import type { DomainErrorCode } from './errors'
import { Lending, type CollectionStatus } from './lending'

describe('Lending.collect', () => {
  const cases: {
    name: string
    amount: number
    collected: number
    collectAmount: number
    wantErr: DomainErrorCode | null
    wantCollected: number
    wantStatus: CollectionStatus
  }[] = [
    {
      name: 'F-1: 一部回収',
      amount: 12_000,
      collected: 0,
      collectAmount: 5_000,
      wantErr: null,
      wantCollected: 5_000,
      wantStatus: 'partial',
    },
    {
      name: 'F-2: 全額に到達',
      amount: 12_000,
      collected: 5_000,
      collectAmount: 7_000,
      wantErr: null,
      wantCollected: 12_000,
      wantStatus: 'collected',
    },
    {
      name: 'F-3: 過回収（一部回収済み）',
      amount: 12_000,
      collected: 5_000,
      collectAmount: 8_000,
      wantErr: 'COLLECT_EXCEEDS_OUTSTANDING',
      wantCollected: 5_000,
      wantStatus: 'partial',
    },
    {
      name: 'F-4: 過回収（回収済み）',
      amount: 12_000,
      collected: 12_000,
      collectAmount: 1,
      wantErr: 'COLLECT_EXCEEDS_OUTSTANDING',
      wantCollected: 12_000,
      wantStatus: 'collected',
    },
    {
      name: 'F-5: 回収額0',
      amount: 12_000,
      collected: 0,
      collectAmount: 0,
      wantErr: 'INVALID_AMOUNT',
      wantCollected: 0,
      wantStatus: 'uncollected',
    },
    {
      name: 'F-6: 回収額が負',
      amount: 12_000,
      collected: 0,
      collectAmount: -100,
      wantErr: 'INVALID_AMOUNT',
      wantCollected: 0,
      wantStatus: 'uncollected',
    },
  ]

  it.each(cases)('$name', (c) => {
    const l = lend(c.amount, c.collected)
    if (c.wantErr === null) {
      l.collect(yen(c.collectAmount))
    } else {
      expectDomainError(() => l.collect(yen(c.collectAmount)), c.wantErr)
    }
    // 失敗時に回収額が動いていないことも見る。最も気づきにくい壊れ方になるため。
    expect(l.collectedAmount).toBe(c.wantCollected)
    expect(l.status()).toBe(c.wantStatus)
  })
})

describe('Lending', () => {
  it('F-7: 未回収なら outstanding は全額', () => {
    const l = lend(12_000, 0)
    expect(l.outstanding()).toBe(12_000)
    expect(l.status()).toBe('uncollected')
    expect(l.isFullyCollected()).toBe(false)
  })

  it('相手が空なら作れない', () => {
    expectDomainError(
      () => Lending.create(id(), '  ', '', yen(12_000), SOME_DATE),
      'EMPTY_COUNTERPARTY',
    )
  })

  it('金額が0以下なら作れない', () => {
    expectDomainError(() => Lending.create(id(), 'テスト相手', '', yen(0), SOME_DATE), 'INVALID_AMOUNT')
  })

  it('作った直後の回収額は0', () => {
    const l = Lending.create(id(), 'テスト相手', 'メモ', yen(12_000), SOME_DATE)
    expect(l.collectedAmount).toBe(0)
    expect(l.status()).toBe('uncollected')
  })

  it('過回収の行は復元できない（CHECK 制約をすり抜けた場合の最後の関門）', () => {
    expectDomainError(
      () => Lending.restore(id(), 'テスト相手', '', yen(12_000), yen(12_001), SOME_DATE),
      'COLLECT_EXCEEDS_OUTSTANDING',
    )
  })
})
